src/test/testserver.rs

```diff
@@ -0,0 +1,54 @@
+use std::io::{self, BufRead, BufReader};
+use std::net::{TcpListener, TcpStream};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::thread;
+
+pub struct TestServer {
+    pub port: u16,
+    pub done: Arc<AtomicBool>,
+}
+
+// Read a stream until reaching a blank line, in order to consume
+// request headers.
+pub fn read_headers(stream: &TcpStream) {
+    for line in BufReader::new(stream).lines() {
+        let line = match line {
+            Ok(x) => x,
+            Err(_) => return,
+        };
```

Comment:
This could be written more succinctly as:

```rust
if line.is_err() { 
    return;
}
```

Or if we want to log or do something with just the error (or value):

```rust
if let Err(e) = line {
   ...
}
```

Reply:
Good point! I just made this an unwrap, since we don't expect errors during the tests and want to panic if we get them.